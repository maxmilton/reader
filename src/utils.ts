export async function getCurrentTab(): Promise<chrome.tabs.Tab> {
  const [tab] = await chrome.tabs.query({ active: true, currentWindow: true });
  return tab;
}

/**
 * Execute a function in the context of the current active tab.
 *
 * Note: The function will be serialised + injected into the tab's page and
 * then run as a content script so it must be self contained (a pure function;
 * no access or mutation outside its own closure scope).
 *
 * @param func - Function to run in page. Must be self-contained.
 * @param args - Arguments to pass to function. Must be JSON-serializable.
 *
 * @see https://developer.chrome.com/docs/extensions/reference/scripting/#method-executeScript
 * @see https://developer.chrome.com/docs/extensions/reference/scripting/#type-ScriptInjection
 * @see https://developer.chrome.com/en/blog/crx-scripting-api/#injecting-a-function-with-arguments
 */
export async function exec<T, A extends any[] = []>(
  func: (..._args: A) => T,
  args?: A,
): Promise<T> {
  const tab = await getCurrentTab();

  const [{ result }] = await chrome.scripting.executeScript({
    target: { tabId: tab.id! },
    func,
    args,
  });

  return result as T;
}

/**
 * Get the Optimal Recognition Point (OPR) index for a given word phrase.
 */
export function indexOfORP(word: string): number {
  let len = word.length;

  while (/\W/.test(word[--len]));

  if (/\W/.test(word[0])) len++;

  switch (++len) {
    case 0:
    case 1: {
      return 0;
    }
    case 2:
    case 3: {
      return 1;
    }
    default: {
      return Math.floor(len / 2) - 1;
    }
  }
}
