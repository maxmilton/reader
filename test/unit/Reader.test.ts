import { afterEach, expect, test } from 'bun:test';
import { cleanup, render } from './utils';

afterEach(cleanup);

// HACK: The Reader component is designed to be rendered once (does not clone
// its view) and mutates global state. Given the global state mutation, it's
// vital to reset its module between tests for accurate test conditions.
const MODULE_PATH = import.meta.resolveSync('../../src/components/Reader');
let Reader: typeof import('../../src/components/Reader').Reader;

async function load(html: string) {
  // @ts-expect-error - stub return value
  global.chrome.scripting.executeScript = () => Promise.resolve([{ result: html }]);

  Loader.registry.delete(MODULE_PATH);
  // eslint-disable-next-line unicorn/no-await-expression-member
  Reader = (await import('../../src/components/Reader')).Reader;
}

// const minimalHTML = '<html><body>x</body></html>';
const basicHTML = await Bun.file('test/unit/fixtures/basic.html').text();
// const brokenHTML = await Bun.file('test/unit/fixtures/broken.html').text();
// const wikipediaSimpleHTML = await Bun.file('test/unit/fixtures/wikipedia-simple.html').text();
// const wikipediaHTML = await Bun.file('test/unit/fixtures/wikipedia.html').text();

// XXX: Because there are async function calls within Reader, we need to wait
// for them to complete before we assert anything.
//  ↳ But not if we want to test the initial state of the DOM. In which case we
//    can cancel any running async tasks.

test('rendered DOM contains expected elements', async () => {
  expect.assertions(16);
  await load(basicHTML);
  const rendered = render(Reader());
  await happyDOM.abort();
  const root = rendered.container.firstChild as HTMLDivElement;
  expect(root).toBeInstanceOf(window.HTMLDivElement);
  expect(rendered.container.querySelector('#progress')).toBeTruthy();
  expect(rendered.container.querySelector('#progress')?.parentNode).toBe(root);
  expect(rendered.container.querySelector('#controls')).toBeTruthy();
  expect(rendered.container.querySelector('#controls')?.parentNode).toBe(root);
  expect(rendered.container.querySelector('#play')).toBeTruthy();
  expect(rendered.container.querySelector('#play')?.parentNode).toBe(
    rendered.container.querySelector('#controls'),
  );
  expect(rendered.container.querySelector('#focus')).toBeTruthy();
  expect(rendered.container.querySelector('#focus')?.parentNode).toBe(root);
  expect(rendered.container.querySelector('#word')).toBeTruthy();
  expect(rendered.container.querySelector('#word')?.parentNode).toBe(root);
  const buttons = rendered.container.querySelectorAll('button');
  expect(buttons).toHaveLength(4);
  expect(buttons[0].textContent).toBe('Rewind');
  expect(buttons[1].textContent).toBe('Play'); // starts as "Play" then changes to "Pause" after load
  expect(buttons[2].textContent).toBe('−');
  expect(buttons[3].textContent).toBe('+');
});

test('rendered DOM initial state matches snapshot', async () => {
  expect.assertions(1);
  await load(basicHTML);
  const rendered = render(Reader());
  await happyDOM.abort();
  expect(rendered.container.innerHTML).toMatchSnapshot();
});

test('rendered DOM playing state matches snapshot', async () => {
  expect.assertions(1);
  await load(basicHTML);
  const rendered = render(Reader());
  await Bun.sleep(1); // lets queued promises in Reader run first
  await happyDOM.abort();
  expect(rendered.container.innerHTML).toMatchSnapshot();
});

test('rendered DOM end state matches snapshot', async () => {
  expect.assertions(1);
  const rendered = render(Reader());
  await happyDOM.waitUntilComplete();
  expect(rendered.container.innerHTML).toMatchSnapshot();
});
