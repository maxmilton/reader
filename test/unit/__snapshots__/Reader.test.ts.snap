// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`rendered DOM initial state matches snapshot 1`] = `"<div><div id="progress"><div id="bar"></div></div><div id="controls"><button>Rewind</button><button id="play">Play</button> <button>−</button><button>+</button></div><div id="focus"></div><div id="word"></div></div>"`;

exports[`rendered DOM playing state matches snapshot 1`] = `"<div><div id="progress"><div id="bar" style="transform: translateX(-88.88888888888889%);"></div></div><div id="controls"><button>Rewind</button><button id="play">Pause</button>180 wpm<button>−</button><button>+</button></div><div id="focus" class="show"></div><div id="word" style="transform: translateX(-0px);"><span id="fp">3</span>.</div></div>"`;

exports[`rendered DOM end state matches snapshot 1`] = `"<div><div id="progress"><div id="bar" style="transform: translateX(0);"></div></div><div id="controls"><button>Rewind</button><button id="play">Play again</button>60000 wpm<button>−</button><button disabled="">+</button></div><div id="focus" class=""></div><div id="word" style=""><div id="summary"><em>ﬁn.</em><br>You read 5 words in 0 seconds.</div></div></div>"`;
